service/s3/api_client.go

```diff
@@ -241,6 +243,21 @@ func addRetryMiddlewares(stack *middleware.Stack, o Options) error {
 	return retry.AddRetryMiddlewares(stack, mo)
 }
 
+// resolves client config
+func resolveClientConfig(cfg aws.Config, o *Options) error {
+	if cfg.ConfigSources == nil || len(cfg.ConfigSources) == 0 {
+		return nil
+	}
```

Comment:
Nit could be simplified to `len(cfg.ConfigSources) == 0` since that returns 0 if the slice is nil.  Or remove the check all together and let `s3sharedconfig.ResolveUseARNRegion` handle it.